### DISCUSSION TITLE
Hook to Bypass Values in AndroidManifest.xml Hook Not Working

### DISCUSSION THREAD

**Discussion by DavidBerdik** [LSPosed/LSPosed#1190] [Q&A]
Hello all,

Yesterday, I posted a question here (https://github.com/LSPosed/LSPosed/discussions/1187) asking for help with fixing an error with writing a module to hook a function. I fixed that issue, so commented on it and marked it as done, however I have a follow-up question: I thought my change from that question would override a value defined in the app's `AndroidManifest.xml`, however, that did not seem to happen, so I am now trying to hook `parseBaseApplication` to do it and it is not working. When I compile this and run it, I don't get any errors. The hook just simply never executes. Does anyone have any ideas as to why or what I can do to fix it?

```java
try {
	XposedHelpers.findAndHookMethod(
			"android.content.pm.PackageParser",
			loadPackageParam.classLoader,
			"parseBaseApplication",
			"android.content.pm.PackageParser.Package",
			Resources.class, XmlResourceParser.class, int.class, String[].class,
			new XC_MethodHook() {
				@Override
				protected void afterHookedMethod(MethodHookParam param) throws Throwable {
					// Get the Package object that was passed as an argument...
					Package pkg = (Package) param.args[0];
					// ...extract the ApplicationInfo object from it.
					ApplicationInfo inf = (ApplicationInfo) XposedHelpers
							.getObjectField(pkg, "applicationInfo");
					// We know the constant for the flag, so apply it to privateFlags...
					final int PRIVATE_FLAG_ALLOW_AUDIO_PLAYBACK_CAPTURE  = 1 << 27;
					int privateFlags = (int) XposedHelpers.getObjectField(inf,
							"privateFlags");
					privateFlags |= PRIVATE_FLAG_ALLOW_AUDIO_PLAYBACK_CAPTURE;
					// Apply the updated privateFlags to ApplicationInfo, then the
					//  updated ApplicationInfo to Package.
					XposedHelpers.setObjectField(inf, "privateFlags", privateFlags);
					XposedHelpers.setObjectField(pkg, "applicationInfo", inf);
					XposedBridge.log("Hooked android.content.pm.parsing." +
							"ParsingPackageImpl#setAllowAudioPlaybackCapture()");
					XposedBridge.log("Hooked android.content.pm.PackageParser" +
							"#parseBaseApplication()");
				}
			}
	);
} catch (Throwable t) {
	XposedBridge.log(t);
}
```

Thank you in advance!

**Comment by DavidBerdik** [comment]
@yujincheng08 Could you please offer me some assistance?

**Comment by YumeMichi** [comment]
@DavidBerdik Sir you fixed this?

**Reply by DavidBerdik** [reply]
Unfortunately, no.

**Comment by buttercookie42** [comment]
I've had the same problem recently – I can't provide a complete solution, either, but two pointers:
- On more recent Android versions (don't ask me when exactly, at some point in the recent years *after* Pie), the Package Parser has been deprecated and at least for internal Framework usage replaced by a new set of classes, e.g. currently [frameworks/base/services/core/java/com/android/server/pm/pkg/parsing/ParsingPackageUtils.java](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/services/core/java/com/android/server/pm/pkg/parsing/ParsingPackageUtils.java;drc=7d3ffbae618e9e728644a96647ed709bf39ae759))
- Likewise I have no idea how things used to work on older Android versions, but currently I have the impression that the manifest might only be read at installation time and then subsequently the parsed values (including any post-processing that might have happened) get cached somewhere, so for overriding on application launch instead of only installation time, you'd have to hook a different set of methods. (I've been looking to override the default max aspect ratio that gets applied to pre-O apps.)

**Reply by DavidBerdik** [reply]
Thank you for sharing this! Even if it's only a partial solution, I appreciate it!

In the two years since I posted this, I ended up delaying the project for which I needed this and have not had time to do further research. When time permits, I want to pick up that project starting with what you have shared here.

**Comment by kdejeqq** [comment]
Good, I'm glad you have this will.💪 The community must not be silent 💪💪

Le jeu. 14 déc. 2023, 19:29, David Berdik ***@***.***> a
écrit :

**Reply by DavidBerdik** [reply]
Well, it might be a little while before I have time to meaningfully dig in to it. Aside from this project, I have a backlog of other Android projects that I'd like to do some work on as well.
